File: catalogo/sistemas/focus-nfe/kit/python/cliente.py
"""Cliente Focus NFe (API v2) — kit do catálogo.

Uso:
    pip install requests
    cp .env.example .env   # preencha FOCUS_AMBIENTE, FOCUS_TOKEN e FOCUS_CNPJ
    python cliente.py                  # só leituras (seguro)
    python cliente.py --emitir-teste   # emite uma NF-e de TESTE (pede confirmação; recusa em produção)

Documentação: https://doc.focusnfe.com.br/reference/introducao
"""
import os
import sys
import time
from pathlib import Path

import requests

BASES = {
    "sandbox": "https://homologacao.focusnfe.com.br",
    "producao": "https://api.focusnfe.com.br",
}


def carregar_env(caminho=Path(__file__).with_name(".env")):
    """Lê um .env simples (CHAVE=valor) sem dependências extras."""
    if not caminho.exists():
        return
    for linha in caminho.read_text(encoding="utf-8").splitlines():
        linha = linha.strip()
        if not linha or linha.startswith("#") or "=" not in linha:
            continue
        chave, valor = linha.split("=", 1)
        os.environ.setdefault(chave.strip(), valor.strip())


class ErroFocus(Exception):
    def __init__(self, status, corpo):
        self.status, self.corpo = status, corpo
        super().__init__(f"Focus NFe HTTP {status}: {corpo}")


class FocusNFe:
    def __init__(self, token=None, ambiente=None, timeout=60):
        carregar_env()
        self.ambiente = (ambiente or os.environ.get("FOCUS_AMBIENTE", "sandbox")).strip()
        if self.ambiente not in BASES:
            raise ValueError("FOCUS_AMBIENTE deve ser 'sandbox' ou 'producao'")
        self.base = BASES[self.ambiente]
        self.token = token or os.environ.get("FOCUS_TOKEN", "")
        if not self.token or self.token.startswith("<<"):
            raise ValueError("Preencha FOCUS_TOKEN no arquivo .env")
        self.timeout = timeout
        self.sessao = requests.Session()
        self.sessao.auth = (self.token, "")  # HTTP Basic: token como usuário, senha vazia

    # ---------------------------------------------------------------- núcleo
    def chamar(self, metodo, caminho, *, params=None, json=None, aceitos=(200,), tentativas=4):
        """Faz a chamada com retry em 429 (Rate-Limit-Reset) e em 5xx/erros de rede."""
        for tentativa in range(tentativas):
            try:
                r = self.sessao.request(metodo, self.base + caminho, params=params, json=json, timeout=self.timeout)
            except requests.RequestException:
                if tentativa == tentativas - 1:
                    raise
                time.sleep(2 ** tentativa)
                continue
            if r.status_code == 429 and tentativa < tentativas - 1:
                time.sleep(int(r.headers.get("Rate-Limit-Reset") or r.headers.get("Retry-After") or 60))
                continue
            if r.status_code >= 500 and tentativa < tentativas - 1:
                time.sleep(2 ** tentativa)
                continue
            if r.status_code not in aceitos:
                corpo = r.json() if "json" in r.headers.get("Content-Type", "") else r.text
                raise ErroFocus(r.status_code, corpo)
            return r
        raise ErroFocus(r.status_code, r.text)

    def _confirmar(self, acao, confirmar):
        if confirmar:
            return
        resp = input(f"[{self.ambiente.upper()} - {self.base}] Confirma {acao}? Digite SIM: ")
        if resp.strip().upper() != "SIM":
            raise RuntimeError("Operação cancelada pelo usuário")

    # ---------------------------------------------------------------- paginação
    def listar_offset(self, caminho, filtros=None):
        """Tabelas auxiliares e empresas: até 50 itens por chamada, total em X-Total-Count."""
        itens, offset = [], 0
        while True:
            r = self.chamar("GET", caminho, params={**(filtros or {}), "offset": offset})
            pagina = r.json()
            if not pagina:
                return itens
            itens.extend(pagina)
            offset += len(pagina)
            total = int(r.headers.get("X-Total-Count", "0") or 0)
            if total and offset >= total:
                return itens

    def listar_versao(self, caminho, filtros, versao=0):
        """Documentos recebidos: versao > X, até 100 por chamada; próximo X em X-Max-Version.
        Retorna (documentos, ultima_versao) — guarde a última versão por CNPJ."""
        docs = []
        while True:
            r = self.chamar("GET", caminho, params={**filtros, "versao": versao})
            pagina = r.json()
            if not pagina:
                return docs, versao
            docs.extend(pagina)
            maxv = r.headers.get("X-Max-Version") or max(int(d.get("versao") or 0) for d in pagina)
            if not maxv or int(maxv) <= versao:
                return docs, versao
            versao = int(maxv)

    # ---------------------------------------------------------------- NF-e
    def emitir_nfe(self, ref, nota, confirmar=False, aguardar=True):
        """POST /v2/nfe?ref=... e, se aguardar, polling até sair de processando_autorizacao."""
        self._confirmar(f"a EMISSÃO da NF-e ref={ref}", confirmar)
        r = self.chamar("POST", "/v2/nfe", params={"ref": ref}, json=nota, aceitos=(201, 202))
        resultado = r.json()
        return self.aguardar("nfe", ref, resultado) if aguardar else resultado

    def aguardar(self, documento, ref, resultado=None, tentativas=10):
        """Consulta GET /v2/<documento>/{ref} com espera crescente (2 s a 60 s)."""
        espera = 2
        resultado = resultado or {"status": "processando_autorizacao"}
        for _ in range(tentativas):
            if resultado.get("status") != "processando_autorizacao":
                break
            time.sleep(espera)
            espera = min(espera * 2, 60)
            resultado = self.chamar("GET", f"/v2/{documento}/{ref}").json()
        return resultado

    def consultar_nfe(self, ref, completa=False):
        return self.chamar("GET", f"/v2/nfe/{ref}", params={"completa": 1 if completa else 0}).json()

    def cancelar_nfe(self, ref, justificativa, confirmar=False):
        if not 15 <= len(justificativa) <= 255:
            raise ValueError("A justificativa deve ter entre 15 e 255 caracteres")
        self._confirmar(f"o CANCELAMENTO (irreversível) da NF-e ref={ref}", confirmar)
        return self.chamar("DELETE", f"/v2/nfe/{ref}", json={"justificativa": justificativa}).json()

    def carta_correcao_nfe(self, ref, correcao, confirmar=False):
        self._confirmar(f"a CARTA DE CORREÇÃO da NF-e ref={ref}", confirmar)
        return self.chamar("POST", f"/v2/nfe/{ref}/carta_correcao", json={"correcao": correcao}).json()

    def inutilizar_nfe(self, cnpj, serie, numero_inicial, numero_final, justificativa, confirmar=False):
        self._confirmar(f"a INUTILIZAÇÃO (irreversível) dos números {numero_inicial}-{numero_final} série {serie}", confirmar)
        corpo = {"cnpj": cnpj, "serie": str(serie), "numero_inicial": str(numero_inicial),
                 "numero_final": str(numero_final), "justificativa": justificativa}
        return self.chamar("POST", "/v2/nfe/inutilizacao", json=corpo).json()

    def baixar_arquivo(self, caminho_relativo, destino):
        """Baixa XML/DANFE a partir de caminho_xml_nota_fiscal, caminho_danfe etc."""
        r = self.chamar("GET", caminho_relativo)
        Path(destino).write_bytes(r.content)
        return destino

    # ---------------------------------------------------------------- NFC-e (síncrona)
    def emitir_nfce(self, ref, nota, confirmar=False):
        self._confirmar(f"a EMISSÃO da NFC-e ref={ref}", confirmar)
        return self.chamar("POST", "/v2/nfce", params={"ref": ref}, json=nota, aceitos=(201,)).json()

    # ---------------------------------------------------------------- NFS-e
    def emitir_nfse(self, ref, nota, confirmar=False, nacional=False, aguardar=True):
        doc = "nfsen" if nacional else "nfse"
        self._confirmar(f"a EMISSÃO da {doc.upper()} ref={ref}", confirmar)
        r = self.chamar("POST", f"/v2/{doc}", params={"ref": ref}, json=nota, aceitos=(201, 202))
        return self.aguardar(doc, ref, r.json()) if aguardar else r.json()

    # ---------------------------------------------------------------- recebidas e auxiliares
    def nfes_recebidas(self, cnpj, versao=0):
        return self.listar_versao("/v2/nfes_recebidas", {"cnpj": cnpj}, versao)

    def manifestar(self, chave, tipo, justificativa=None, confirmar=False):
        self._confirmar(f"a MANIFESTAÇÃO '{tipo}' da NF-e {chave}", confirmar)
        corpo = {"tipo": tipo, **({"justificativa": justificativa} if justificativa else {})}
        return self.chamar("POST", f"/v2/nfes_recebidas/{chave}/manifesto", json=corpo).json()

    def cadastrar_webhook(self, cnpj, evento, url, segredo=None, cabecalho=None, confirmar=False):
        self._confirmar(f"o cadastro do webhook '{evento}' para {url}", confirmar)
        corpo = {"cnpj": cnpj, "event": evento, "url": url}
        if segredo:
            corpo.update({"authorization": segredo, "authorization_header": cabecalho or "Authorization"})
        return self.chamar("POST", "/v2/hooks", json=corpo).json()

    def cep(self, cep):
        return self.chamar("GET", f"/v2/ceps/{cep}").json()

    def ncm(self, codigo):
        return self.chamar("GET", f"/v2/ncms/{codigo}").json()

    def municipio(self, codigo_ibge):
        return self.chamar("GET", f"/v2/municipios/{codigo_ibge}").json()


NOTA_TESTE = {
    "natureza_operacao": "Venda de mercadoria",
    "data_emissao": time.strftime("%Y-%m-%dT%H:%M:%S-03:00"),
    "tipo_documento": 1, "finalidade_emissao": 1, "local_destino": 1,
    "consumidor_final": 1, "presenca_comprador": 2,
    # texto que a SEFAZ costuma exigir no destinatário em homologação
    "nome_destinatario": "NF-E EMITIDA EM AMBIENTE DE HOMOLOGACAO - SEM VALOR FISCAL",
    "cpf_destinatario": "00000000000", "indicador_inscricao_estadual_destinatario": 9,
    "logradouro_destinatario": "Rua Exemplo", "numero_destinatario": "100", "bairro_destinatario": "Centro",
    "municipio_destinatario": "São Paulo", "uf_destinatario": "SP", "cep_destinatario": "01001000",
    "modalidade_frete": 9, "valor_produtos": 10.0, "valor_total": 10.0,
    "items": [{
        "numero_item": 1, "codigo_produto": "TESTE-1", "descricao": "Produto de teste", "cfop": "5102",
        "codigo_ncm": "61091000", "unidade_comercial": "UN", "quantidade_comercial": 1,
        "valor_unitario_comercial": 10.0, "unidade_tributavel": "UN", "quantidade_tributavel": 1,
        "valor_unitario_tributavel": 10.0, "valor_bruto": 10.0, "icms_origem": 0,
        "icms_situacao_tributaria": "102", "pis_situacao_tributaria": "07", "cofins_situacao_tributaria": "07",
    }],
    "formas_pagamento": [{"forma_pagamento": "01", "valor_pagamento": 10.0}],
}


if __name__ == "__main__":
    api = FocusNFe()
    cnpj = os.environ.get("FOCUS_CNPJ", "")
    print(f"Ambiente: {api.ambiente} ({api.base})")

    # Leituras (seguras)
    print("CEP 01001000:", api.cep("01001000"))
    if cnpj and not cnpj.startswith("<<"):
        docs, versao = api.nfes_recebidas(cnpj)
        print(f"{len(docs)} NF-e recebidas; guarde a versão {versao} para a próxima carga")

    # Escrita de teste (somente sandbox e com confirmação)
    if "--emitir-teste" in sys.argv:
        if api.ambiente != "sandbox":
            sys.exit("A emissão de teste só roda com FOCUS_AMBIENTE=sandbox")
        nota = dict(NOTA_TESTE, cnpj_emitente=cnpj)
        ref = "teste" + time.strftime("%Y%m%d%H%M%S")
        try:
            resultado = api.emitir_nfe(ref, nota)
            print("Resultado:", resultado.get("status"), resultado.get("mensagem_sefaz"), resultado.get("erros"))
            if resultado.get("status") == "autorizado":
                arq = api.baixar_arquivo(resultado["caminho_xml_nota_fiscal"], f"{ref}.xml")
                print("XML salvo em", arq)
        except ErroFocus as e:
            print("Erro da API:", e.status, e.corpo)

File: catalogo/sistemas/focus-nfe/kit/typescript/cliente.ts
/**
 * Cliente Focus NFe (API v2) — kit do catálogo. Node 18+ (fetch nativo).
 *
 *   cp .env.example .env    # preencha FOCUS_AMBIENTE, FOCUS_TOKEN e FOCUS_CNPJ
 *   node --env-file=.env --experimental-strip-types cliente.ts     # Node 22.6+
 *   npx tsx --env-file=.env cliente.ts                             # alternativa
 *   ... cliente.ts --emitir-teste   # emite NF-e de TESTE (só sandbox, pede confirmação)
 *
 * Documentação: https://doc.focusnfe.com.br/reference/introducao
 */
import { writeFile } from "node:fs/promises";
import { createInterface } from "node:readline/promises";

type Ambiente = "sandbox" | "producao";
type Json = Record<string, any>;

const BASES: Record<Ambiente, string> = {
  sandbox: "https://homologacao.focusnfe.com.br",
  producao: "https://api.focusnfe.com.br",
};

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export class ErroFocus extends Error {
  constructor(public status: number, public corpo: unknown) {
    super(`Focus NFe HTTP ${status}: ${typeof corpo === "string" ? corpo : JSON.stringify(corpo)}`);
  }
}

export class FocusNFe {
  readonly ambiente: Ambiente;
  readonly base: string;
  private readonly auth: string;

  constructor(token = process.env.FOCUS_TOKEN ?? "", ambiente = (process.env.FOCUS_AMBIENTE ?? "sandbox") as Ambiente) {
    if (!(ambiente in BASES)) throw new Error("FOCUS_AMBIENTE deve ser 'sandbox' ou 'producao'");
    if (!token || token.startsWith("<<")) throw new Error("Preencha FOCUS_TOKEN no arquivo .env");
    this.ambiente = ambiente;
    this.base = BASES[ambiente];
    // HTTP Basic: token como usuário, senha vazia
    this.auth = "Basic " + Buffer.from(`${token}:`).toString("base64");
  }

  /** Chamada com retry em 429 (Rate-Limit-Reset) e 5xx. */
  async chamar(metodo: string, caminho: string, opts: { query?: Record<string, string | number>; corpo?: unknown; aceitos?: number[] } = {}): Promise<Response> {
    const aceitos = opts.aceitos ?? [200];
    const qs = opts.query ? "?" + new URLSearchParams(Object.entries(opts.query).map(([k, v]) => [k, String(v)])) : "";
    for (let tentativa = 0; ; tentativa++) {
      const resp = await fetch(this.base + caminho + qs, {
        method: metodo,
        headers: { Authorization: this.auth, "Content-Type": "application/json", Accept: "application/json" },
        body: opts.corpo === undefined ? undefined : JSON.stringify(opts.corpo),
      });
      if (resp.status === 429 && tentativa < 3) {
        await sleep(Number(resp.headers.get("Rate-Limit-Reset") ?? resp.headers.get("Retry-After") ?? "60") * 1000);
        continue;
      }
      if (resp.status >= 500 && tentativa < 3) {
        await sleep(2 ** tentativa * 1000);
        continue;
      }
      if (!aceitos.includes(resp.status)) {
        const texto = await resp.text();
        let corpo: unknown = texto;
        try { corpo = JSON.parse(texto); } catch { /* 401 vem em text/html */ }
        throw new ErroFocus(resp.status, corpo);
      }
      return resp;
    }
  }

  private async confirmar(acao: string, confirmado: boolean): Promise<void> {
    if (confirmado) return;
    const rl = createInterface({ input: process.stdin, output: process.stdout });
    const r = await rl.question(`[${this.ambiente.toUpperCase()} - ${this.base}] Confirma ${acao}? Digite SIM: `);
    rl.close();
    if (r.trim().toUpperCase() !== "SIM") throw new Error("Operação cancelada pelo usuário");
  }

  /** Tabelas auxiliares e empresas: 50 por página, total em X-Total-Count. */
  async listarOffset(caminho: string, filtros: Record<string, string> = {}): Promise<Json[]> {
    const itens: Json[] = [];
    let offset = 0;
    while (true) {
      const resp = await this.chamar("GET", caminho, { query: { ...filtros, offset } });
      const pagina = (await resp.json()) as Json[];
      if (pagina.length === 0) return itens;
      itens.push(...pagina);
      offset += pagina.length;
      const total = Number(resp.headers.get("X-Total-Count") ?? 0);
      if (total && offset >= total) return itens;
    }
  }

  /** Documentos recebidos: versao > X (até 100 por chamada); próximo X em X-Max-Version. */
  async listarVersao(caminho: string, filtros: Record<string, string>, versao = 0): Promise<{ docs: Json[]; versao: number }> {
    const docs: Json[] = [];
    while (true) {
      const resp = await this.chamar("GET", caminho, { query: { ...filtros, versao } });
      const pagina = (await resp.json()) as Json[];
      if (pagina.length === 0) return { docs, versao };
      docs.push(...pagina);
      const maxV = Number(resp.headers.get("X-Max-Version") ?? Math.max(...pagina.map((d) => Number(d.versao ?? 0))));
      if (!maxV || maxV <= versao) return { docs, versao };
      versao = maxV;
    }
  }

  /** Polling de GET /v2/<documento>/{ref} com espera crescente (2 s a 60 s). */
  async aguardar(documento: string, ref: string, resultado: Json = { status: "processando_autorizacao" }, tentativas = 10): Promise<Json> {
    let espera = 2000;
    for (let i = 0; i < tentativas && resultado.status === "processando_autorizacao"; i++) {
      await sleep(espera);
      espera = Math.min(espera * 2, 60000);
      resultado = (await (await this.chamar("GET", `/v2/${documento}/${ref}`)).json()) as Json;
    }
    return resultado;
  }

  async emitirNfe(ref: string, nota: Json, confirmado = false, aguardar = true): Promise<Json> {
    await this.confirmar(`a EMISSÃO da NF-e ref=${ref}`, confirmado);
    const resp = await this.chamar("POST", "/v2/nfe", { query: { ref }, corpo: nota, aceitos: [201, 202] });
    const r = (await resp.json()) as Json;
    return aguardar ? this.aguardar("nfe", ref, r) : r;
  }

  async consultarNfe(ref: string, completa = false): Promise<Json> {
    return (await this.chamar("GET", `/v2/nfe/${ref}`, { query: { completa: completa ? 1 : 0 } })).json();
  }

  async cancelarNfe(ref: string, justificativa: string, confirmado = false): Promise<Json> {
    if (justificativa.length < 15 || justificativa.length > 255) throw new Error("Justificativa deve ter 15 a 255 caracteres");
    await this.confirmar(`o CANCELAMENTO (irreversível) da NF-e ref=${ref}`, confirmado);
    return (await this.chamar("DELETE", `/v2/nfe/${ref}`, { corpo: { justificativa } })).json();
  }

  async cartaCorrecaoNfe(ref: string, correcao: string, confirmado = false): Promise<Json> {
    await this.confirmar(`a CARTA DE CORREÇÃO da NF-e ref=${ref}`, confirmado);
    return (await this.chamar("POST", `/v2/nfe/${ref}/carta_correcao`, { corpo: { correcao } })).json();
  }

  async inutilizarNfe(cnpj: string, serie: string, numeroInicial: string, numeroFinal: string, justificativa: string, confirmado = false): Promise<Json> {
    await this.confirmar(`a INUTILIZAÇÃO (irreversível) de ${numeroInicial}-${numeroFinal} série ${serie}`, confirmado);
    const corpo = { cnpj, serie, numero_inicial: numeroInicial, numero_final: numeroFinal, justificativa };
    return (await this.chamar("POST", "/v2/nfe/inutilizacao", { corpo })).json();
  }

  async emitirNfce(ref: string, nota: Json, confirmado = false): Promise<Json> {
    await this.confirmar(`a EMISSÃO da NFC-e ref=${ref}`, confirmado);
    return (await this.chamar("POST", "/v2/nfce", { query: { ref }, corpo: nota, aceitos: [201] })).json();
  }

  async emitirNfse(ref: string, nota: Json, nacional = false, confirmado = false): Promise<Json> {
    const doc = nacional ? "nfsen" : "nfse";
    await this.confirmar(`a EMISSÃO da ${doc.toUpperCase()} ref=${ref}`, confirmado);
    const r = (await (await this.chamar("POST", `/v2/${doc}`, { query: { ref }, corpo: nota, aceitos: [201, 202] })).json()) as Json;
    return this.aguardar(doc, ref, r);
  }

  /** Baixa XML/DANFE a partir de caminho_xml_nota_fiscal, caminho_danfe etc. */
  async baixarArquivo(caminhoRelativo: string, destino: string): Promise<string> {
    const resp = await this.chamar("GET", caminhoRelativo);
    await writeFile(destino, Buffer.from(await resp.arrayBuffer()));
    return destino;
  }

  nfesRecebidas(cnpj: string, versao = 0) {
    return this.listarVersao("/v2/nfes_recebidas", { cnpj }, versao);
  }

  async manifestar(chave: string, tipo: "ciencia" | "confirmacao" | "desconhecimento" | "nao_realizada", justificativa?: string, confirmado = false): Promise<Json> {
    await this.confirmar(`a MANIFESTAÇÃO '${tipo}' da NF-e ${chave}`, confirmado);
    return (await this.chamar("POST", `/v2/nfes_recebidas/${chave}/manifesto`, { corpo: { tipo, ...(justificativa ? { justificativa } : {}) } })).json();
  }

  async cep(cep: string): Promise<Json> {
    return (await this.chamar("GET", `/v2/ceps/${cep}`)).json();
  }
}

const NOTA_TESTE: Json = {
  natureza_operacao: "Venda de mercadoria",
  tipo_documento: 1, finalidade_emissao: 1, local_destino: 1, consumidor_final: 1, presenca_comprador: 2,
  // Texto que a SEFAZ costuma exigir no destinatário em homologação
  nome_destinatario: "NF-E EMITIDA EM AMBIENTE DE HOMOLOGACAO - SEM VALOR FISCAL",
  cpf_destinatario: "00000000000", indicador_inscricao_estadual_destinatario: 9,
  logradouro_destinatario: "Rua Exemplo", numero_destinatario: "100", bairro_destinatario: "Centro",
  municipio_destinatario: "São Paulo", uf_destinatario: "SP", cep_destinatario: "01001000",
  modalidade_frete: 9, valor_produtos: 10, valor_total: 10,
  items: [{
    numero_item: 1, codigo_produto: "TESTE-1", descricao: "Produto de teste", cfop: "5102", codigo_ncm: "61091000",
    unidade_comercial: "UN", quantidade_comercial: 1, valor_unitario_comercial: 10, unidade_tributavel: "UN",
    quantidade_tributavel: 1, valor_unitario_tributavel: 10, valor_bruto: 10, icms_origem: 0,
    icms_situacao_tributaria: "102", pis_situacao_tributaria: "07", cofins_situacao_tributaria: "07",
  }],
  formas_pagamento: [{ forma_pagamento: "01", valor_pagamento: 10 }],
};

async function main() {
  const api = new FocusNFe();
  const cnpj = process.env.FOCUS_CNPJ ?? "";
  console.log(`Ambiente: ${api.ambiente} (${api.base})`);
  console.log("CEP 01001000:", await api.cep("01001000"));
  if (cnpj && !cnpj.startsWith("<<")) {
    const { docs, versao } = await api.nfesRecebidas(cnpj);
    console.log(`${docs.length} NF-e recebidas; guarde a versão ${versao}`);
  }
  if (process.argv.includes("--emitir-teste")) {
    if (api.ambiente !== "sandbox") throw new Error("A emissão de teste só roda com FOCUS_AMBIENTE=sandbox");
    const ref = "teste" + Date.now();
    const nota = { ...NOTA_TESTE, cnpj_emitente: cnpj, data_emissao: new Date().toISOString() };
    const r = await api.emitirNfe(ref, nota);
    console.log("Resultado:", r.status, r.mensagem_sefaz ?? "", r.erros ?? "");
    if (r.status === "autorizado") console.log("XML salvo em", await api.baixarArquivo(r.caminho_xml_nota_fiscal, `${ref}.xml`));
  }
}

main().catch((e) => {
  console.error(e instanceof ErroFocus ? `Erro da API: ${e.status} ${JSON.stringify(e.corpo)}` : e);
  process.exit(1);
});
